lighthouse-core/test/gather/computed/main-thread-tasks-test.js

```diff
@@ -0,0 +1,128 @@
+/**
+ * @license Copyright 2017 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* eslint-env mocha */
+
+const Runner = require('../../../runner.js');
+const taskGroups = require('../../../lib/task-groups.js').taskGroups;
+const pwaTrace = require('../../fixtures/traces/progressive-app.json');
+const TracingProcessor = require('../../../lib/traces/tracing-processor.js');
+const assert = require('assert');
+
+describe('MainResource computed artifact', () => {
+  const args = {data: {}};
+  const baseTs = 1241250325;
+  let boilerplateTrace;
+  let computedArtifacts;
+
+  beforeEach(() => {
+    boilerplateTrace = [
+      {ph: 'I', name: 'TracingStartedInPage', ts: baseTs, args},
+      {ph: 'I', name: 'navigationStart', ts: baseTs, args},
+    ];
+    computedArtifacts = Runner.instantiateComputedArtifacts();
+  });
+
+  it('should get all main thread tasks from a trace', async () => {
+    const tasks = await computedArtifacts.requestMainThreadTasks({traceEvents: pwaTrace});
+    const toplevelTasks = tasks.filter(task => !task.parent);
+    assert.equal(tasks.length, 2305);
+    assert.equal(toplevelTasks.length, 296);
+
+    // Sanity check the reachability of tasks and summation of selfTime
+    const allTasks = [];
+    const queue = toplevelTasks;
+    let totalTime = 0;
+    let totalTopLevelTime = 0;
+    while (queue.length) {
+      const task = queue.shift();
+      totalTime += task.selfTime;
+      totalTopLevelTime += TracingProcessor.isScheduleableTask(task.event) ? task.duration : 0;
+      allTasks.push(task);
+      queue.push(...task.children);
+    }
+
+    assert.equal(allTasks.length, 2305);
+    assert.equal(Math.round(totalTopLevelTime), 386);
+    assert.equal(Math.round(totalTime), 396);
+  });
+
+  it('should compute parent/child correctly', async () => {
+    const traceEvents = [
```

Comment:
here i made this for you:

```js
    /*
    An artistic rendering of the below trace:
    █████████████████████████████TaskA██████████████████████████████████████████████
          ████████████████TaskB███████████████████
               ███████TaskC██████████
    */
```

i also made an x-axis but it scale correctly against the box-drawing characters in any monospace fonts except the one i personally use. :( 

```
    |----|-----|---|---|------------------|-----------------|--------------------------------------------|-
    0    5     10  15  20                 40                55                                           100 
```